<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="CAMeRa">
<title>Development notes • CAMeRa</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Development notes">
<meta property="og:description" content="CAMeRa">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">CAMeRa</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/development.html">Development notes</a>
    <a class="dropdown-item" href="../articles/import-local.html">Import to CAMERA from local data</a>
    <a class="dropdown-item" href="../articles/tutorial.html">Tutorial</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/universe77/CAMERA">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Development notes</h1>
            
      
      
      <div class="d-none name"><code>development.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>CAMERA (Cross Ancestral MEndelian RAndomization) is a method to
estimate the causal effect of an exposure on an outcome using GWAS
summary statistics from different populations.The CAMERA package
includes three important components:</p>
<ul>
<li>Instrument selection for multi-ancestral Mendelian randomization
(MR) analysis</li>
<li>Causal inference across multi-ancestral populations using structural
equation model (SEM)</li>
<li>Identification of pleiotropic effects across multiple ancestral
populations</li>
</ul>
<p>Each step is documented in the documentation. The following analyses
should run within a couple of minutes, depending on internet speed and
the traffic that the <a href="https://gwas.mrcieu.ac.uk/" class="external-link">IEU GWAS
database</a> servers are experiencing.</p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>To install CAMERA:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"universe77/CAMERA"</span><span class="op">)</span></span></code></pre></div>
<p>And then load it into your R session</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CAMeRa</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data-setup">Data setup<a class="anchor" aria-label="anchor" href="#data-setup"></a>
</h3>
<p>CAMERA begins by choosing an exposure and outcome hypothesis that can
be tested in multi-ancestral populations. Here, we will be estimating
the causal effect of body mass index (BMI) on coronary heart disease
(CHD) in European ancestry (EUR) and East Asian ancestry (EAS).</p>
<p>Summary statistics data can be extracted from the IEU GWAS database
using the <a href="https://mrcieu.github.io/TwoSampleMR/" class="external-link">TwoSampleMR</a> package. A
list of available traits can be obtained using:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traits</span> <span class="op">&lt;-</span> <span class="fu">TwoSampleMR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/available_outcomes.html" class="external-link">available_outcomes</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>You can also browse the available traits here: <a href="https://gwas.mrcieu.ac.uk/" class="external-link uri">https://gwas.mrcieu.ac.uk/</a></p>
<p>Once you obtain the study IDs for the exposure and the outcome, open
R6 class environment to run CAMERA. The minimum information required for
CAMERA is the following:</p>
<ul>
<li>Summary statistics for the exposure and the outcome</li>
<li>Population information</li>
<li>Plink (ver.1.90)</li>
<li>LD reference data</li>
</ul>
<p>Plink (ver.1.90) and LD reference data are required to identify
instruments that can be used for both populations. LD reference data can
be accessed from: <a href="http://fileserve.mrcieu.ac.uk/ld/1kg.v3.tgz" class="external-link uri">http://fileserve.mrcieu.ac.uk/ld/1kg.v3.tgz</a></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Need to specify path to plink and LD reference data here.</span></span>
<span><span class="co"># The following exposures and the outcomes will be used: Exposures: BMI (EUR, EAS); Outcomes: CHD (EUR, EAS)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"genetics.binaRies"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Package \"genetics.binaRies\" must be installed to run this chunk."</span><span class="op">)</span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/knit_exit.html" class="external-link">knit_exit</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/CAMERA.html">CAMERA</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>    exposure_ids<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ieu-a-2"</span>, <span class="st">"bbj-a-1"</span><span class="op">)</span>, </span>
<span>    outcome_ids<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ieu-a-7"</span>, <span class="st">"bbj-a-109"</span><span class="op">)</span>, </span>
<span>    pops <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EUR"</span>, <span class="st">"EAS"</span><span class="op">)</span>,</span>
<span>    bfiles<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"/Users/yc16575/OneDrive - University of Bristol/EUR"</span>, <span class="st">"/Users/yc16575/OneDrive - University of Bristol/EAS"</span><span class="op">)</span>,         </span>
<span>    plink <span class="op">=</span> <span class="fu">genetics.binaRies</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/genetics.binaRies/man/get_plink_binary.html" class="external-link">get_plink_binary</a></span><span class="op">(</span><span class="op">)</span>,      </span>
<span>    radius<span class="op">=</span><span class="fl">50000</span>, </span>
<span>    clump_pop<span class="op">=</span><span class="st">"EAS"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Make sure that the exposures/outcomes are matched across the
populations. Population 1 and 2 should have the same exposure-outcome
pair. Also, instrument-trait associations should be consistent between
the populations (e.g. how similar SNP-BMI association in EUR to SNP-BMI
assocation in EAS). You can check this as follows:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">check_phenotypes</span><span class="op">(</span>ids<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">exposure_ids</span><span class="op">)</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">check_phenotypes</span><span class="op">(</span>ids<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">outcome_ids</span><span class="op">)</span></span></code></pre></div>
<p>A value closes to 1 indicates that the instrument-trait associations
are consistent between the populations. If the values is too low or too
high, it is recommended to use other available instruments for the
traits.</p>
</div>
<div class="section level3">
<h3 id="running-camera">Running CAMERA<a class="anchor" aria-label="anchor" href="#running-camera"></a>
</h3>
<p>We can now perform the analysis, which will do the following:</p>
<ol style="list-style-type: decimal">
<li>Extract instruments for the exposures</li>
<li>Check the validity of the instruments across the populations
(Standardise/scale the data if necessary)</li>
<li>Extract new instruments based on LD information and
fine-mapping</li>
<li>Extract instruments for the outcomes</li>
<li>Harmonise the exposure data and the outcome data</li>
<li>Perform MR using SEM</li>
</ol>
</div>
<div class="section level3">
<h3 id="step-by-step-analysis">Step by step analysis<a class="anchor" aria-label="anchor" href="#step-by-step-analysis"></a>
</h3>
<p>See the <code><a href="../reference/CAMERA.html">?CAMERA</a></code> for options on the parameters for this
analysis.</p>
<div class="section level4">
<h4 id="step-1--extract-instruments-for-the-exposures-from-each-pouplation">Step 1. Extract instruments for the exposures from each
pouplation<a class="anchor" aria-label="anchor" href="#step-1--extract-instruments-for-the-exposures-from-each-pouplation"></a>
</h4>
<p>This returns a set of LD clumped SNPs that are GWAS significant for
BMI (raw instruments). The following function identifies SNPs that have
strong associations with the exposure in each population. This is the
same method as instrument extraction for multivariable MR. A data frame
of the extracted instruments is stored in
<code>x$instrument_raw</code></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">extract_instruments</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_raw</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-2--check-the-validity-of-the-instruments">Step 2. Check the validity of the instruments<a class="anchor" aria-label="anchor" href="#step-2--check-the-validity-of-the-instruments"></a>
</h4>
<p>It is important to ensure that the instruments for the exposure are
valid across the populations. Once instruments for the exposure trait
are identified for each population, we can assess specificity of the
instruments. Each of the following functions estimates heterogeneity of
the instruments between and calculates fraction of the instruments
(obtained from the Step 1) that are replicated between the
populations.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">instrument_heterogeneity</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_raw</span><span class="op">)</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">estimate_instrument_specificity</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_raw</span>, alpha <span class="op">=</span> <span class="st">"bonferroni"</span><span class="op">)</span></span></code></pre></div>
<p>To correct for winner’s curse:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">estimate_instrument_specificity</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_raw</span>, alpha <span class="op">=</span> <span class="st">"bonferroni"</span>, winnerscurse<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-3--extract-new-instruments-based-on-ld-information-and-fine-mapping">Step 3. Extract new instruments based on LD information and
fine-mapping<a class="anchor" aria-label="anchor" href="#step-3--extract-new-instruments-based-on-ld-information-and-fine-mapping"></a>
</h4>
<p>For Cross/Multi-ancestral MR it is important to ensure that genetic
structures of the populations are properly considered. The following
functions will identify a new set of SNPs that have the largest effect
on the exposure, considering LD patterns. For the cases where the most
strongly associated SNPs are not available at exactly the same position,
the function searches alternative SNPs that are located near the
original SNP and show the largest effect size magnitude (MaxZ method). A
data frame of the extracted instruments is stored in
<code>x$instrument_maxz</code></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify genomic regions that include the raw instruments (x$instrument_raw)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="fu">extract_instrument_regions</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Extract instrument from identified region by searching across region for best associations</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">scan_regional_instruments</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># Plot regions and their region-selected instruments</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">plot_regional_instruments</span><span class="op">(</span>instruments<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_maxz</span><span class="op">)</span></span></code></pre></div>
<p>Check the replication fraction between the populations
(instrument_maxz).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Estimate what fraction of the region-selected instruments we'd expect to replicate</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">instrument_heterogeneity</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_raw</span><span class="op">)</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">estimate_instrument_specificity</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_raw</span>, alpha <span class="op">=</span> <span class="st">"bonferroni"</span>, winnerscurse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Also, instruments can be obtained using fine mapping method.
Extracted instruments are stored in <code>x$instrument_susie</code> or
<code>x$instrument_susie</code></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate LD matrix for the whole regions that include the raw instruments</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">regional_ld_matrices</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># Run Susie</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">susie_finemap_regions</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># Run PAINTOR</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">paintor_finemap_regions</span><span class="op">(</span>PAINTOR<span class="op">=</span><span class="st">"/user/work/yc16575/PAINTOR_V3.0/PAINTOR"</span>, workdir<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-4--extract-instruments-for-the-outcomes">Step 4. Extract instruments for the outcomes<a class="anchor" aria-label="anchor" href="#step-4--extract-instruments-for-the-outcomes"></a>
</h4>
<p>Once instruments for the exposure trait have been specified, those
variants need to be extracted from the outcome trait. You can also
specify p-value thresholds.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get outcome data for a particular set of instruments and run SEM model </span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">make_outcome_data</span><span class="op">(</span>exp<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_raw</span>, p_exp<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_outcome</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-5--harmonise-the-exposure-data-and-the-outcome-data">Step 5. Harmonise the exposure data and the outcome data<a class="anchor" aria-label="anchor" href="#step-5--harmonise-the-exposure-data-and-the-outcome-data"></a>
</h4>
<p>This creates a new data frame that has the exposure data and outcome
data combined (e.g. exposure-outcome for EUR and EAS).</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">harmonised_dat</span><span class="op">(</span>exp<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_raw</span>, out<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_outcome</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">harmonised_dat_sem</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-6--perform-mr-using-genomic-structural-equation-model-mr-sem">Step 6. Perform MR using genomic structural equation model
(MR-SEM)<a class="anchor" aria-label="anchor" href="#step-6--perform-mr-using-genomic-structural-equation-model-mr-sem"></a>
</h4>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">perform_basic_sem</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>You can also perform MR-SEM using the instruments from the Step
3.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Standardise the betas and SEs if required</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">make_outcome_data</span><span class="op">(</span>exp<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_maxz</span>, p_exp<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_outcome</span><span class="op">)</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">harmonised_dat</span><span class="op">(</span>exp<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_maxz</span>, out<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_outcome</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">harmonised_dat_sem</span><span class="op">)</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">perform_basic_sem</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="standardisation-scaling">Standardisation / scaling<a class="anchor" aria-label="anchor" href="#standardisation-scaling"></a>
</h4>
<p>If the units for the exposures/the outcomes are different across the
populations, run:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">standardise_data</span><span class="op">(</span>dat<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_raw</span>, standardise_unit<span class="op">=</span><span class="cn">TRUE</span>, standardise_scale<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">standardised_instrument_raw</span><span class="op">)</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">standardise_data</span><span class="op">(</span>dat<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_outcome</span>, standardise_unit<span class="op">=</span><span class="cn">TRUE</span>, standardise_scale<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">standardised_instrument_outcome</span><span class="op">)</span></span></code></pre></div>
<p>Note that standardisation is not required if the effect size of
SNP-trait is in the unit of SD or log odds.</p>
<p>For the cases where the SNP-exposure or SNP-outcome association shows
substantial difference between the population, scale the beta
estimates.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">standardise_data</span><span class="op">(</span>dat<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_raw</span>, standardise_unit<span class="op">=</span><span class="cn">TRUE</span>, standardise_scale<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">standardised_instrument_raw</span><span class="op">)</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">standardise_data</span><span class="op">(</span>dat<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">instrument_outcome</span>, standardise_unit<span class="op">=</span><span class="cn">TRUE</span>, standardise_scale<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">standardised_outcome</span><span class="op">)</span></span></code></pre></div>
<p>You can check specificity of the instruments again using:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">instrument_heterogeneity</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">standardised_instrument_raw</span><span class="op">)</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">estimate_instrument_specificity</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">standardised_instrument_raw</span>, alpha <span class="op">=</span> <span class="st">"bonferroni"</span>, winnerscurse<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>You can also make harmonised data again after standaridisation /
scaling:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">harmonised_dat</span><span class="op">(</span>exp<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">standardised_instrument_raw</span>, out<span class="op">=</span><span class="va">x</span><span class="op">$</span><span class="va">standardised_outcome</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">harmonised_dat_sem</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yoonsu Cho, Gibran Hemani.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
